piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+        //let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        //assert_eq!(input.graphemes(true).count(), 1);
+        //assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point_x as f64), Some(layout.width()));
+        //assert_eq!(input.len(), 17);
+
+        let input = "\u{0023}\u{FE0F}\u{20E3}"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(layout.width()));
+        assert_eq!(input.len(), 7);
+        assert_eq!(input.chars().count(), 3);
+    }
+
+    #[test]
+    fn test_hit_test_text_position_complex_1() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        // Notes on this input:
+        // 5 code points
+        // 5 utf-16 code units
+        // 10 utf-8 code units (2/1/3/3/1)
+        // 3 graphemes
+        let input = "é\u{0023}\u{FE0F}\u{20E3}1"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        let test_layout_0 = text_layout.new_text_layout(&font, "é").build().unwrap();
+        let test_layout_1 = text_layout.new_text_layout(&font, "é\u{0023}\u{FE0F}\u{20E3}").build().unwrap();
+
+        //assert_eq!(input.graphemes(true).count(), 3);
+        assert_eq!(input.len(), 10);
+
+        // Note: text position is in terms of utf8 code units
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(test_layout_0.width()));
+        assert_eq!(layout.hit_test_text_position(2, true).map(|p| p.point.x), Some(test_layout_1.width()));
+        assert_eq!(layout.hit_test_text_position(9, true).map(|p| p.point.x), Some(layout.width()));
+
+        assert_eq!(layout.hit_test_text_position(2, false).map(|p| p.point.x), Some(test_layout_0.width()));
+        assert_eq!(layout.hit_test_text_position(9, false).map(|p| p.point.x), Some(test_layout_1.width()));
+    }
+
+    #[test]
+    fn test_hit_test_point_basic() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        let mut text_layout = D2DText::new(&dwrite);
+
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, "piet text!").build().unwrap();
+        println!("text pos 4 leading: {:?}", layout.hit_test_text_position(4, false)); // 20.302734375
+        println!("text pos 4 trailing: {:?}", layout.hit_test_text_position(4, true)); // 23.58984375
+
+        // test hit test point
+        // all inside
+        let pt = layout.hit_test_point(Point::new(21.0, 0.0));
```

Comment:
Urgh. Using metrics from the font is fragile. Otoh, there's a built-in "slop" here because of the snap-to-closest semantics. So I'm ok with this staying as-is and fixing if/when tests fail. I'm also curious to run it on Windows 7.

Based on my reading of the tests, it implements "round to nearest" behavior, then `is_trailing_hit` indicates whether the rounding was up or down.